File: Graph Algorithms/Diijkstra_Algorithm.py

def find_lowest_cost_node(costs):
	lowest_cost = float("inf")
	lowest_cost_node = None
	for node in costs:
		cost = costs[node]
		if cost < lowest_cost and node not in processed:
			lowest_cost = cost
			lowest_cost_node = node

	return lowest_cost_node

graph = {}
graph["start"] = {}
graph["start"]["a"] = 5
graph["start"]["b"] = 2

graph["a"] = {}
graph["a"]["c"] = 4
graph["a"]["d"] = 2

graph["b"] = {}
graph["b"]["a"] = 8
graph["b"]["d"] = 7

graph["c"] = {}
graph["c"]["d"] = 6
graph["c"]["fim"] = 3

graph["d"] = {}
graph["d"]["fim"] = 1

graph["fim"] = {}

infinity = float("inf")

costs = {}
costs["a"] = 5
costs["b"] = 2
costs["c"] = infinity
costs["d"] = infinity
costs["fim"] = infinity

parents = {}
parents["a"] = "start"
parents["b"] = "start"
parents["c"] = "a"
parents["d"] = "b"
parents["fim"] = None

processed = []

node = find_lowest_cost_node(costs)  
while node is not None:
	print("node: " + node)
	cost = costs[node]
	print("cost: ",cost)
	neighbors = graph[node]
	for n in neighbors.keys():
		new_cost = cost + neighbors[n]
		if costs[n] > new_cost:
			costs[n] = new_cost
			parents[n] = node

	processed.append(node)
	node = find_lowest_cost_node(costs)

print(costs["fim"])

File: Graph Algorithms/Diijkstra_Algorithm.cpp
#include <bits/stdc++.h>
using namespace std;

string find_lowest_cost_node(map <string, int> costs, set <string> processed)
{
	int lowest_cost = 1000000000;
	string lowest_cost_node;

	map<string, int>::iterator it = costs.begin();
	for (;it != costs.end(); it++)
	{
		int cost = costs[it->first]; 
		if (cost < lowest_cost && processed.find(it->first) == processed.end() )
		{
			lowest_cost = cost;
			lowest_cost_node.clear();
			lowest_cost_node = it->first; 
		}
	}
	
	return lowest_cost_node;
}


int main(int argc, char const *argv[])
{
	map < string, map<string, int> > graph;
 	map <string, int> costs;
 	map <string, string> parents;
	set <string> processed;
	int infinity = INT_MAX;

	graph["start"]["a"] = 6;
	graph["start"]["b"] = 2; 
	graph["a"]["fim"] = 1; 
	graph["b"]["a"] = 3;
	graph["b"]["fim"] = 5;
	graph["fim"]["fim"] = 0;
	costs["a"] = 6;
	costs["b"] = 2;
	costs["fim"] = infinity;
	parents["a"] = "start";
	parents["b"] = "start";
	parents["fim"] = "NULL";
	
	string node = find_lowest_cost_node(costs,processed);
	while (!node.empty())
	{
		int cost = costs[node];
		map<string, int> neighbors (graph[node].begin(), graph[node].end());
		map<string, int>::iterator it = neighbors.begin();
		for (;it != neighbors.end(); it++)
		{
			int new_cost = cost + neighbors[it->first];
			if (costs[it->first] > new_cost)
			{
				costs[it->first] = new_cost;
				parents[it->first] = node;
			}
		}

		processed.insert(node);
		node = find_lowest_cost_node(costs,processed);
	}

	cout << costs["fim"] << endl;

	return 0;
}